In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib
matplotlib.rcParams["figure.figsize"]=(20,10)

## business objective

I want to conduct a research on the effects of two variants of covid 19 i.e. delta and omicorn ,
assuming the samples collected by me are all the patients either affected by delta or omicorn 
In this project im only cosidering who have been recovered from this disease  withing considering the patients who had already died due to this disease

# Generating the dataset

In [10]:
no_days_before_recover=pd.Series(np.random.randint(1,30,20))
no_days_before_recover

0     12
1      3
2     11
3     21
4      1
5     15
6      3
7     11
8     15
9      7
10     7
11    11
12    24
13     8
14    25
15    28
16    26
17    22
18    12
19    15
dtype: int32

In [11]:
cat1=pd.Series(np.random.choice(["Male","Female"],20))
cat1

0     Female
1     Female
2     Female
3     Female
4     Female
5       Male
6       Male
7     Female
8       Male
9       Male
10      Male
11      Male
12    Female
13      Male
14      Male
15      Male
16    Female
17    Female
18    Female
19      Male
dtype: object

In [12]:
cat2=pd.Series(np.random.choice(["india","china","usa","italy"],20))
cat2

0     india
1       usa
2     china
3     italy
4     italy
5     italy
6       usa
7     china
8       usa
9     india
10    italy
11    india
12      usa
13    india
14    italy
15    india
16    china
17      usa
18      usa
19    italy
dtype: object

In [13]:
age=pd.Series(np.random.randint(10,80,20))
age

0     38
1     32
2     51
3     19
4     45
5     60
6     24
7     49
8     70
9     24
10    42
11    13
12    52
13    65
14    47
15    49
16    63
17    57
18    69
19    15
dtype: int32

In [14]:
cat3=pd.Series(np.random.choice(["yes","no"],20))
cat3

0     yes
1      no
2     yes
3     yes
4     yes
5      no
6      no
7      no
8      no
9      no
10    yes
11    yes
12     no
13     no
14     no
15    yes
16     no
17     no
18     no
19     no
dtype: object

In [15]:
cat4=pd.Series(np.random.choice(["delta","omicorn"],20))
cat4

0     omicorn
1     omicorn
2       delta
3     omicorn
4     omicorn
5       delta
6     omicorn
7     omicorn
8       delta
9       delta
10      delta
11      delta
12    omicorn
13      delta
14    omicorn
15      delta
16      delta
17    omicorn
18    omicorn
19    omicorn
dtype: object

In [17]:
df=pd.concat([cat1,cat2,age,cat3,cat4,no_days_before_recover],axis=1)
df.columns=["Gender","country","age","has_comorbidity","variant","no_days_before_recover"]

In [18]:
df

Gender country  age has_comorbidity  variant  no_days_before_recover
0   Female   india   38             yes  omicorn                      12
1   Female     usa   32              no  omicorn                       3
2   Female   china   51             yes    delta                      11
3   Female   italy   19             yes  omicorn                      21
4   Female   italy   45             yes  omicorn                       1
5     Male   italy   60              no    delta                      15
6     Male     usa   24              no  omicorn                       3
7   Female   china   49              no  omicorn                      11
8     Male     usa   70              no    delta                      15
9     Male   india   24              no    delta                       7
10    Male   italy   42             yes    delta                       7
11    Male   india   13             yes    delta                      11
12  Female     usa   52              no  omicorn                      24
13    Male   india   65              no    delta                       8
14    Male   italy   47              no  omicorn                      25
15    Male   india   49             yes    delta                      28
16  Female   china   63              no    delta                      26
17  Female     usa   57              no  omicorn                      22
18  Female     usa   69              no  omicorn                      12
19    Male   italy   15              no  omicorn                      15

In [72]:
tab1=pd.crosstab(df.country,df.variant,margins=True)
tab1

variant  delta  omicorn  All
country                     
china        2        1    3
india        4        1    5
italy        2        4    6
usa          1        5    6
All          9       11   20

# Z Test for Proportion

In [73]:
from statsmodels.stats.proportion import proportions_ztest

In [74]:
tab1.loc["All":]

variant  delta  omicorn  All
country                     
All          9       11   20

In [75]:
cal_tab=tab1.loc[["india","china"],["delta","All"]]
cal_tab

variant  delta  All
country            
india        4    5
china        2    3

In [78]:
def prop_test_registered(c):
    print(c)
    stat, pval = proportions_ztest(cal_tab["delta"],cal_tab["All"])
    print(stat)
    print(pval)
    print("  ")

In [79]:
prop_test_registered(cal_tab)

variant  delta  All
country            
india        4    5
china        2    3
0.4216370213557842
0.6732899796599954
  


## p value is 0.6732. Hence there is no significant difference in propertions of delta in india and china

# Packages for statistical analysis

In [26]:
import statsmodels.api as sm
from scipy.stats import ttest_ind,chi2_contingency
from statsmodels.stats.weightstats import ztest
from statsmodels.stats.proportion import proportions_ztest

# Anova

In [63]:
from scipy.stats import f_oneway

In [64]:
a=[]

In [65]:
g=df.groupby("country")["no_days_before_recover"]


In [66]:
for i in g:
    a.append(i[1].values)

In [67]:
a

[array([11, 11, 26]),
 array([12,  7, 11,  8, 28]),
 array([21,  1, 15,  7, 25, 15]),
 array([ 3,  3, 15, 24, 22, 12])]

In [68]:
f_oneway(a[0],a[1],a[2],a[3])

F_onewayResult(statistic=0.08151759181951218, pvalue=0.9691252448518685)

## The p value is 0.9691.Hence there is no  significant difference in no of days before recover between the country

# TTEST

In [49]:
def ttest(col):
    print(f"FOR {col}")
    delta_no_days_before_recover=df[df.variant=="delta"]["no_days_before_recover"]
    omicorn_no_days_before_recover=df[df.variant=="omicorn"]["no_days_before_recover"]
    print(delta_no_days_before_recover.values)
    print(omicorn_no_days_before_recover.values)
    print(ttest_ind(delta_no_days_before_recover.values,omicorn_no_days_before_recover.values))

In [50]:
ttest("variant")

FOR variant
[11 15 15  7  7 11  8 28 26]
[12  3 21  1  3 11 24 25 22 12 15]
Ttest_indResult(statistic=0.1803768181989022, pvalue=0.8588720034473295)


## p value is 0.8588.Hence there is no significant difference in no_days_before_recover between delta and omicorn

# CHI SQUARE TEST

In [46]:
def chi_square(crosstab):
    stat, p, dof, expected = chi2_contingency(crosstab)
    return(p)

In [47]:
ct=pd.crosstab(df["country"],df["variant"])
print(ct)

variant  delta  omicorn
country                
china        2        1
india        4        1
italy        2        4
usa          1        5


In [48]:
print(chi_square(ct))

0.14981854206745496


## p value is 0.1498.Hence the country and variant(omicorn and delta) have a significant association between them 

# Multiple Regression

In [42]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
import statsmodels.api as sm

In [43]:
df

Gender country  age has_comorbidity  variant  no_days_before_recover
0   Female   india   38             yes  omicorn                      12
1   Female     usa   32              no  omicorn                       3
2   Female   china   51             yes    delta                      11
3   Female   italy   19             yes  omicorn                      21
4   Female   italy   45             yes  omicorn                       1
5     Male   italy   60              no    delta                      15
6     Male     usa   24              no  omicorn                       3
7   Female   china   49              no  omicorn                      11
8     Male     usa   70              no    delta                      15
9     Male   india   24              no    delta                       7
10    Male   italy   42             yes    delta                       7
11    Male   india   13             yes    delta                      11
12  Female     usa   52              no  omicorn                      24
13    Male   india   65              no    delta                       8
14    Male   italy   47              no  omicorn                      25
15    Male   india   49             yes    delta                      28
16  Female   china   63              no    delta                      26
17  Female     usa   57              no  omicorn                      22
18  Female     usa   69              no  omicorn                      12
19    Male   italy   15              no  omicorn                      15

In [44]:
X_cat=df[["Gender","country","has_comorbidity","variant"]]
X_cat

Gender country has_comorbidity  variant
0   Female   india             yes  omicorn
1   Female     usa              no  omicorn
2   Female   china             yes    delta
3   Female   italy             yes  omicorn
4   Female   italy             yes  omicorn
5     Male   italy              no    delta
6     Male     usa              no  omicorn
7   Female   china              no  omicorn
8     Male     usa              no    delta
9     Male   india              no    delta
10    Male   italy             yes    delta
11    Male   india             yes    delta
12  Female     usa              no  omicorn
13    Male   india              no    delta
14    Male   italy              no  omicorn
15    Male   india             yes    delta
16  Female   china              no    delta
17  Female     usa              no  omicorn
18  Female     usa              no  omicorn
19    Male   italy              no  omicorn

In [45]:
cat=pd.get_dummies(X_cat,drop_first=True)
cat

Gender_Male  country_india  country_italy  country_usa  \
0             0              1              0            0   
1             0              0              0            1   
2             0              0              0            0   
3             0              0              1            0   
4             0              0              1            0   
5             1              0              1            0   
6             1              0              0            1   
7             0              0              0            0   
8             1              0              0            1   
9             1              1              0            0   
10            1              0              1            0   
11            1              1              0            0   
12            0              0              0            1   
13            1              1              0            0   
14            1              0              1            0   
15            1              1              0            0   
16            0              0              0            0   
17            0              0              0            1   
18            0              0              0            1   
19            1              0              1            0   

    has_comorbidity_yes  variant_omicorn  
0                     1                1  
1                     0                1  
2                     1                0  
3                     1                1  
4                     1                1  
5                     0                0  
6                     0                1  
7                     0                1  
8                     0                0  
9                     0                0  
10                    1                0  
11                    1                0  
12                    0                1  
13                    0                0  
14                    0                1  
15                    1                0  
16                    0                0  
17                    0                1  
18                    0                1  
19                    0                1

In [80]:
cont=df[["age"]]
cont

age
0    38
1    32
2    51
3    19
4    45
5    60
6    24
7    49
8    70
9    24
10   42
11   13
12   52
13   65
14   47
15   49
16   63
17   57
18   69
19   15

In [81]:
cont_scaled=pd.DataFrame(StandardScaler().fit_transform(cont),columns=cont.columns)

In [82]:
cont_scaled

age
0  -0.354686
1  -0.697930
2   0.389010
3  -1.441625
4   0.045766
5   0.903876
6  -1.155588
7   0.274595
8   1.475949
9  -1.155588
10 -0.125856
11 -1.784869
12  0.446217
13  1.189913
14  0.160181
15  0.274595
16  1.075498
17  0.732254
18  1.418742
19 -1.670454

In [83]:
X=pd.concat([cont_scaled,cat],axis=1)

In [84]:
y=df["no_days_before_recover"]

In [85]:
def regression_results(X,y):
    x=sm.add_constant(X)
    model=sm.OLS(y,x).fit()
    print(model.summary())

In [86]:
regression_results(X,y)

                              OLS Regression Results                              
Dep. Variable:     no_days_before_recover   R-squared:                       0.096
Model:                                OLS   Adj. R-squared:                 -0.431
Method:                     Least Squares   F-statistic:                    0.1822
Date:                    Wed, 03 Aug 2022   Prob (F-statistic):              0.984
Time:                            12:12:47   Log-Likelihood:                -68.773
No. Observations:                      20   AIC:                             153.5
Df Residuals:                          12   BIC:                             161.5
Df Model:                               7                                         
Covariance Type:                nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------

## The p value associated with F statistic is 0.1822.Hence this model is not significant.

## R2 is 0.096. Adj R2 is -0.431. 

## all P values are greater than 0.05.Hence none of the predictors are significant in determining the no_of_days_to recover of the patients.